File: trunk/source/jugglinglab/notation/mhnHands.java
// mhnHands.java
//
// Copyright 2004 by Jack Boyce (jboyce@users.sourceforge.net) and others

/*
    This file is part of Juggling Lab.

    Juggling Lab is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Juggling Lab is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Juggling Lab; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

package jugglinglab.notation;

import java.util.*;
import java.text.MessageFormat;

import jugglinglab.util.*;


public class mhnHands {
    // static ResourceBundle guistrings;
    static ResourceBundle errorstrings;
    static {
        // guistrings = JLLocale.getBundle("GUIStrings");
        errorstrings = JLLocale.getBundle("ErrorStrings");
    }

    protected int jugglers = 0;
    protected int[] size = null;
    protected int[][] coords = null;
    protected int[][] catches = null;
    protected double[][][][] handpath = null;
    

    public mhnHands(String hands) throws JuggleExceptionUser, JuggleExceptionInternal {
        // delete the '<' and '>' characters first
        int pos;
        while ((pos = hands.indexOf('<')) >= 0) {
            hands = hands.substring(0,pos) + hands.substring(pos+1,hands.length());
        }
        while ((pos = hands.indexOf('>')) >= 0) {
            hands = hands.substring(0,pos) + hands.substring(pos+1,hands.length());
        }
        while ((pos = hands.indexOf('{')) >= 0) {
            hands = hands.substring(0,pos) + hands.substring(pos+1,hands.length());
        }
        while ((pos = hands.indexOf('}')) >= 0) {
            hands = hands.substring(0,pos) + hands.substring(pos+1,hands.length());
        }

        StringTokenizer st1 = new StringTokenizer(hands, "|!", false);
        jugglers = st1.countTokens();

        size = new int[jugglers];
        coords = new int[jugglers][];
        catches = new int[jugglers][];
        handpath = new double[jugglers][][][];

        for (int j = 0; j < jugglers; j++) {
            String str = st1.nextToken();
            // System.out.println("str["+j+"] = "+str);

            for (int k = 0; k < 3; k++) {
                pos = 0;
                int numcoords = 0;
                boolean gotthrow = false, gotcatch = false;

                for (int l = 0; l < str.length(); ) {
                    char ch = str.charAt(l);

                    if (ch == ' ') {
                        l++;
                        continue;
                    }
                    if (ch == '.') {
                        if (k == 1) {
                            coords[j][pos] = numcoords;
                            if (coords[j][pos] < 2)
                                throw new JuggleExceptionUser(errorstrings.getString("Error_hands_toofewcoords"));

                            handpath[j][pos] = new double[numcoords][];
                        } else if (k == 2) {
                            if (!gotcatch)
                                catches[j][pos] = coords[j][pos] - 1;
                            if (handpath[j][pos][0] == null)
                                throw new JuggleExceptionUser(errorstrings.getString("Error_hands_nothrow"));
                            if (handpath[j][pos][catches[j][pos]] == null)
                                throw new JuggleExceptionUser(errorstrings.getString("Error_hands_nocatch"));
                        }
                        gotthrow = gotcatch = false;
                        pos++;
                        numcoords = 0;
                        l++;
                        continue;
                    }
                    if (ch == '-') {
                        if (k == 2)
                            handpath[j][pos][numcoords] = null;
                        numcoords++;
                        l++;
                        continue;
                    }
                    if ((ch == 'T') || (ch == 't')) {
                        if (numcoords != 0)
                            throw new JuggleExceptionUser(errorstrings.getString("Error_hands_Tnotstart"));
                        if (gotthrow)
                            throw new JuggleExceptionUser(errorstrings.getString("Error_hands_toomanycoords"));
                        gotthrow = true;
                        l++;
                        continue;
                    }
                    if ((ch == 'C') || (ch == 'c')) {
                        if (numcoords == 0)
                            throw new JuggleExceptionUser(errorstrings.getString("Error_hands_Catstart"));
                        if (gotcatch)
                            throw new JuggleExceptionUser(errorstrings.getString("Error_hands_toomanycatches"));
                        if (k == 1)
                            catches[j][pos] = numcoords;
                        gotcatch = true;
                        l++;
                        continue;
                    }
                    if (ch == '(') {
                        int endindex = str.indexOf(')', l+1);
                        if (endindex < 0)
                            throw new JuggleExceptionUser(errorstrings.getString("Error_hands_noparen"));
                        if (k == 2) {
                            handpath[j][pos][numcoords] = new double[3];
                            String str2 = str.substring(l+1, endindex);

                            try {
                                StringTokenizer st4 = new StringTokenizer(str2, ",", false);
                                handpath[j][pos][numcoords][0] =
                                    Double.valueOf(st4.nextToken()).doubleValue();
                                if (st4.hasMoreTokens())
                                    handpath[j][pos][numcoords][2] =
                                        Double.valueOf(st4.nextToken()).doubleValue();
                                if (st4.hasMoreTokens())
                                    handpath[j][pos][numcoords][1] =
                                        Double.valueOf(st4.nextToken()).doubleValue();
                            } catch (NumberFormatException e) {
                                throw new JuggleExceptionUser(errorstrings.getString("Error_hands_coordinate"));
                            } catch (NoSuchElementException e) {
                                throw new JuggleExceptionInternal("No such element exception in \"hands\"");
                            }
                        }
                        numcoords++;
                        l = endindex + 1;
                        continue;
                    }

					String template = errorstrings.getString("Error_hands_character");
					Object[] arguments = { Character.toString(ch) };					
					throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
                }

                if (k == 0) {
                    size[j] = pos;
                    catches[j] = new int[pos];
                    coords[j] = new int[pos];
                    handpath[j] = new double[pos][][];
                }

                if (numcoords != 0)
                    throw new JuggleExceptionUser(errorstrings.getString("Error_hands_badending"));
            }
        }

    }

    public int getPeriod(int juggler) {
        int j = (juggler - 1) % jugglers;
        return size[j];
    }

    public int getNumberOfCoordinates(int juggler, int pos) {
        int j = (juggler - 1) % jugglers;
        return coords[j][pos];
    }

    // throw index is always 0:
    public int getCatchIndex(int juggler, int pos) {
        int j = (juggler - 1) % jugglers;
        return catches[j][pos];
    }

    // both pos and index are indexed from 0:
    public Coordinate getCoordinate(int juggler, int pos, int index) {
        if ((pos >= getPeriod(juggler)) || (index >= getNumberOfCoordinates(juggler, pos)))
            return null;
        int j = (juggler - 1) % jugglers;
        if (handpath[j][pos][index] == null)
            return null;
        return new Coordinate(handpath[j][pos][index][0], handpath[j][pos][index][1],
                              handpath[j][pos][index][2]);
    }
}

File: composeApp/src/commonMain/kotlin/jugglinglab/notation/MHNHands.kt
//
// MHNHands.kt
//
// This class parses the "hands" parameter in MHN notation.
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.notation

import jugglinglab.composeapp.generated.resources.*
import jugglinglab.util.*

data class MHNHands(
    val str: String
) {
    val numberOfJugglers: Int
    private val size: IntArray
    private val coords: Array<IntArray>
    private val catches: Array<IntArray>
    private val handpath: Array<Array<Array<DoubleArray?>>>

    init {
        val cleanStr = jlExpandRepeats(str).filterNot { it in "<>{}" }
        val jugglerStrings = cleanStr.split('|', '!')

        numberOfJugglers = jugglerStrings.size

        val jugglerData = jugglerStrings.map { jugglerStr ->
            val beatStrings = jlSplitOnCharOutsideParens(jugglerStr.trim(), '.')
            beatStrings.map { beatStr ->
                val coordTokens = mutableListOf<DoubleArray?>()
                var catchIndex = -1
                var gotThrow = false

                var pos = 0
                while (pos < beatStr.length) {
                    when (val ch = beatStr[pos]) {
                        ' ' -> {
                            // character is ignored
                            ++pos
                        }
                        '-' -> {
                            // placeholder; interpolate from other nearby coordinates
                            coordTokens.add(null)
                            ++pos
                        }
                        'T', 't' -> {
                            // marks a throw
                            if (coordTokens.isNotEmpty()) {
                                val message = getStringResource(Res.string.error_hands_tnotstart)
                                throw JuggleExceptionUser(message)
                            }
                            if (gotThrow) {
                                val message = getStringResource(Res.string.error_hands_toomanycoords)
                                throw JuggleExceptionUser(message)
                            }
                            gotThrow = true
                            ++pos
                        }
                        'C', 'c' -> {
                            // marks a catch
                            if (coordTokens.isEmpty()) {
                                val message = getStringResource(Res.string.error_hands_catstart)
                                throw JuggleExceptionUser(message)
                            }
                            if (catchIndex != -1) {
                                val message = getStringResource(Res.string.error_hands_toomanycatches)
                                throw JuggleExceptionUser(message)
                            }
                            catchIndex = coordTokens.size
                            ++pos
                        }
                        '(' -> {
                            // position coordinate specified
                            val closeIndex = beatStr.indexOf(')', pos + 1)
                            if (closeIndex < 0) {
                                val message = getStringResource(Res.string.error_hands_noparen)
                                throw JuggleExceptionUser(message)
                            }

                            val coordStr = beatStr.substring(pos + 1, closeIndex)
                            try {
                                val parts = coordStr.split(',').map { jlParseFiniteDouble(it.trim()) }
                                val coord = DoubleArray(3)
                                coord[0] = parts.getOrElse(0) { 0.0 }
                                // Note: y and z are swapped from input
                                coord[2] = parts.getOrElse(1) { 0.0 }
                                coord[1] = parts.getOrElse(2) { 0.0 }
                                coordTokens.add(coord)
                            } catch (_: NumberFormatException) {
                                val message = getStringResource(Res.string.error_hands_coordinate)
                                throw JuggleExceptionUser(message)
                            }
                            pos = closeIndex + 1
                        }
                        else -> {
                            val message = getStringResource(Res.string.error_hands_character, ch.toString())
                            throw JuggleExceptionUser(message)
                        }
                    }
                }

                if (coordTokens.size < 2) {
                    val message = getStringResource(Res.string.error_hands_toofewcoords)
                    throw JuggleExceptionUser(message)
                }
                if (coordTokens[0] == null) {
                    val message = getStringResource(Res.string.error_hands_nothrow)
                    throw JuggleExceptionUser(message)
                }

                val finalCatchIndex = if (catchIndex != -1) catchIndex else coordTokens.size - 1
                if (coordTokens[finalCatchIndex] == null) {
                    val message = getStringResource(Res.string.error_hands_nocatch)
                    throw JuggleExceptionUser(message)
                }

                Triple(coordTokens.size, finalCatchIndex, coordTokens.toTypedArray())
            }
        }
        size = IntArray(numberOfJugglers) { jugglerData[it].size }
        coords = Array(numberOfJugglers) { jugglerIndex ->
            IntArray(jugglerData[jugglerIndex].size) { beatIndex ->
                jugglerData[jugglerIndex][beatIndex].first
            }
        }
        catches = Array(numberOfJugglers) { jugglerIndex ->
            IntArray(jugglerData[jugglerIndex].size) { beatIndex ->
                jugglerData[jugglerIndex][beatIndex].second
            }
        }
        handpath = Array(numberOfJugglers) { jugglerIndex ->
            Array(jugglerData[jugglerIndex].size) { beatIndex ->
                jugglerData[jugglerIndex][beatIndex].third
            }
        }
    }

    fun getPeriod(juggler: Int): Int {
        val j = (juggler - 1) % numberOfJugglers
        return size[j]
    }

    fun getNumberOfCoordinates(juggler: Int, pos: Int): Int {
        val j = (juggler - 1) % numberOfJugglers
        return coords[j][pos]
    }

    // Return the index value for juggler `juggler` and beat `pos`, when the
    // catch is made.
    //
    // Note the throw is always assumed to happen at index 0.

    fun getCatchIndex(juggler: Int, pos: Int): Int {
        val j = (juggler - 1) % numberOfJugglers
        return catches[j][pos]
    }

    // Return the hand coordinate at `index`, for juggler `juggler` on beat `pos`.
    // Note that both `pos` and `index` are indexed from 0.

    fun getCoordinate(juggler: Int, pos: Int, index: Int): Coordinate? {
        if (pos >= getPeriod(juggler) || index >= getNumberOfCoordinates(juggler, pos)) {
            return null
        }
        val j = (juggler - 1) % numberOfJugglers
        return if (handpath[j][pos][index] == null) {
            null
        } else {
            Coordinate(
                handpath[j][pos][index]!![0],
                handpath[j][pos][index]!![1],
                handpath[j][pos][index]!![2]
            )
        }
    }
}
